File: guides/ai-github-developer-py/ai_github_developer/actions.py
import os
import random
import string
from typing import Any, Dict
from e2b import Sandbox
from rich.console import Console
from rich.theme import Theme

REPO_DIRECTORY = "/home/user/repo"

custom_theme = Theme(
    {
        "sandbox_action": "bold #E57B00",  # Adjust color as needed
    }
)

console = Console(theme=custom_theme)


def print_sandbox_action(action_type: str, action_message: str):
    console.print(
        f"[sandbox_action] [Sandbox Action][/sandbox_action] {action_type}: {action_message}"
    )


# List of actions for the assistant
def create_directory(sandbox: Sandbox, args: Dict[str, Any]) -> str:
    directory = args["path"]
    print_sandbox_action("Creating directory", directory)

    try:
        sandbox.filesystem.make_dir(directory)
        return "success"
    except Exception as e:
        return f"Error: {e}"


def save_content_to_file(sandbox: Sandbox, args: Dict[str, Any]) -> str:
    path = args["path"]
    content = args["content"]
    print_sandbox_action("Saving content to", path)

    try:
        dir = os.path.dirname(path)
        sandbox.filesystem.make_dir(dir)
        sandbox.filesystem.write(path, content)
        return "success"
    except Exception as e:
        return f"Error: {e}"


def list_files(sandbox: Sandbox, args: Dict[str, Any]) -> str:
    path = args["path"]
    print_sandbox_action("Listing files on path", path)

    try:
        files = sandbox.filesystem.list(path)
        response = "\n".join(
            [f"dir: {file.name}" if file.is_dir else file.name for file in files]
        )
        return response
    except Exception as e:
        return f"Error: {e}"


def read_file(sandbox: Sandbox, args: Dict[str, Any]) -> str:
    path = args["path"]
    print_sandbox_action("Reading file on path", path)

    try:
        return sandbox.filesystem.read(path)
    except Exception as e:
        return f"Error: {e}"


def commit(sandbox: Sandbox, args: Dict[str, Any]) -> str:
    repo_directory = "/home/user/repo"
    commit_message = args["message"]
    print_sandbox_action("Committing with the message", commit_message)

    git_add_proc = sandbox.process.start_and_wait(f"git -C {repo_directory} add .")
    if git_add_proc.exit_code != 0:
        error = f"Error adding files to staging: {git_add_proc.stdout}\n\t{git_add_proc.stderr}"
        console.print("\t[bold red]Error:[/bold red]", error)
        return error

    git_commit_proc = sandbox.process.start_and_wait(
        f"git -C {repo_directory} commit -m '{commit_message}'"
    )
    if git_commit_proc.exit_code != 0:
        error = f"Error committing changes: {git_commit_proc.stdout}\n\t{git_commit_proc.stderr}"
        console.print("\t[bold red]Error:[/bold red]", error)
        return error

    return "success"


def make_pull_request(sandbox: Sandbox, args: Dict[str, Any]) -> str:
    base_branch = "main"
    random_letters = "".join(random.choice(string.ascii_letters) for _ in range(5))
    new_branch_name = f"ai-developer-{random_letters}"

    title = args["title"]
    body = ""

    print_sandbox_action(
        "Making a pull request", f"from '{new_branch_name}' to '{base_branch}'"
    )

    git_checkout_proc = sandbox.process.start_and_wait(
        f"git -C {REPO_DIRECTORY} checkout -b {new_branch_name}"
    )
    if git_checkout_proc.exit_code != 0:
        error = f"Error creating a new git branch {new_branch_name}: {git_checkout_proc.stdout}\n\t{git_checkout_proc.stderr}"
        console.print("\t[bold red]Error:[/bold red]", error)
        return error

    git_push_proc = sandbox.process.start_and_wait(
        f"git -C {REPO_DIRECTORY} push -u origin {new_branch_name}"
    )
    if git_push_proc.exit_code != 0:
        error = (
            f"Error pushing changes: {git_push_proc.stdout}\n\t{git_push_proc.stderr}"
        )
        console.print("\t[bold red]Error:[/bold red]", error)
        return error

    gh_pull_request_proc = sandbox.process.start_and_wait(
        cmd=f'gh pr create --base "{base_branch}" --head "{new_branch_name}" --title "{title}" --body "{body}"'.replace(
            "`", "\\`"
        ),
        cwd=REPO_DIRECTORY,
    )
    if gh_pull_request_proc.exit_code != 0:
        error = f"Error creating pull request: {gh_pull_request_proc.stdout}\n\t{gh_pull_request_proc.stderr}"
        console.print("\t[bold red]Error:[/bold red]", error)
        return error

    return "success"

File: guides/ai-github-developer-js/actions.ts
import {Sandbox} from '@e2b/sdk';
import path from 'path';


export const REPO_DIRECTORY = '/home/user/repo';

export async function createDirectory(
    sandbox: Sandbox,
    { directory }: { directory: string },
): Promise<string> {
    console.log("Creating directory", directory)
    try {
        await sandbox.filesystem.makeDir(directory);
        return 'success';
    } catch (e) {
        // @ts-ignore
        const error = `Error: ${e.message}`;
        console.error(error);
        return error;
    }
}
export async function saveContentToFile(
    sandbox: Sandbox,
    { content, filePath }: { content: string; filePath: string },
): Promise<string> {
    console.log("Saving content to", filePath)
    try {
        const dir = path.dirname(filePath);

        await sandbox.filesystem.makeDir(dir);
        await sandbox.filesystem.write(filePath, content);

        return 'success';
    } catch (e) {
        // @ts-ignore
        const error = `Error: ${e.message}`;
        console.error(error);
        return error;
    }
}

export async function listFiles(
    sandbox: Sandbox,
    { dirPath }: { dirPath: string }, // Updated parameter name to avoid conflict with reserved word
): Promise<string> {
    console.log("Listing files on path", dirPath)
    try {
        const files = await sandbox.filesystem.list(dirPath); // Updated parameter name
        return files.map((file) =>
            file.isDir ? `dir: ${file.name}` : file.name
        ).join('\n');
    } catch (e) {
        // @ts-ignore
        const error = `Error: ${e.message}`;
        console.error(error);
        return error;
    }
}

export async function readFile(
    sandbox: Sandbox,
    { filePath }: { filePath: string }, // Updated parameter name
): Promise<string> {
    console.log("Reading file on path", filePath)
    try {
        return await sandbox.filesystem.read(filePath); // Updated parameter name
    } catch (e) {
        // @ts-ignore
        const error = `Error: ${e.message}`;
        console.error(error);
        return error;
    }
}

export async function commit(
    sandbox: Sandbox,
    { commitMessage }: { commitMessage: string },
): Promise<string> {
    console.log("Committing with the message", commitMessage)

    try {
        const gitAddProc = await sandbox.process.startAndWait(`git -C ${REPO_DIRECTORY} add .`);
        if (gitAddProc.exitCode != 0) {
            const error = `Error adding files to staging: ${gitAddProc.stdout}\n\t${gitAddProc.stderr}`
            console.error(error)
            return error
        }


        const commitProc = await sandbox.process.startAndWait(
            `git -C ${REPO_DIRECTORY} commit -m "${commitMessage}"`,
        );

        if (commitProc.exitCode != 0) {
            const error = `Error committing changes: ${commitProc.stdout}\n\t${commitProc.stderr}`
            console.error(error)
            return error
        }

        return 'success';
    } catch (e) {
        // @ts-ignore
        const error = `Error: ${e.message}`;
        console.error(error);
        return error;
    }
}

export async function makePullRequest(
    sandbox: Sandbox,
    args: { [arg: string]: any },
): Promise<string> {
    const REPO_DIRECTORY = '/home/user/repo';

    const baseBranch = 'main';
    const randomLetters = Array.from({ length: 5 }, () =>
        String.fromCharCode(65 + Math.floor(Math.random() * 26))
    ).join('');
    const newBranchName = `ai-developer-${randomLetters}`;

    const title = args.title;
    const body = '';

    try {
        console.log(
            'Making a pull request',
            `from '${newBranchName}' to '${baseBranch}'`
        );

        const gitCheckoutProc = await sandbox.process.startAndWait(
            `git -C ${REPO_DIRECTORY} checkout -b ${newBranchName}`
        );
        if (gitCheckoutProc.exitCode !== 0) {
            const error = `Error creating a new git branch ${newBranchName}: ${gitCheckoutProc.stdout}\n\t${gitCheckoutProc.stderr}`;
            console.error(error);
            return error;
        }

        const gitPushProc = await sandbox.process.startAndWait(
            `git -C ${REPO_DIRECTORY} push -u origin ${newBranchName}`
        );
        if (gitPushProc.exitCode !== 0) {
            const error = `Error pushing changes: ${gitPushProc.stdout}\n\t${gitPushProc.stderr}`;
            console.error(error);
            return error;
        }

        sandbox.cwd = REPO_DIRECTORY;
        const ghPullRequestProc = await sandbox.process.startAndWait(
            `gh pr create --base "${baseBranch}" --head "${newBranchName}" --title "${title}" --body "${body.replace(
                /"/g,
                '\\"'
            )}"`,
        );
        if (ghPullRequestProc.exitCode !== 0) {
            const error = `Error creating pull request: ${ghPullRequestProc.stdout}\n\t${ghPullRequestProc.stderr}`;
            console.error(error);
            return error;
        }

        return 'success';
    } catch (e) {
        // @ts-ignore
        const error = `Error: ${e.message}`;
        console.error(error);
        return error;
    }
}
